src/conversions/channels.rs

```diff
@@ -106,36 +113,62 @@ mod test {
 
     #[test]
     fn remove_channels() {
-        let input = vec![1u16, 2, 3, 1, 2, 3];
+        let input = vec![1u16, 2, 3, 4, 5, 6];
         let output = ChannelCountConverter::new(input.into_iter(), 3, 2).collect::<Vec<_>>();
-        assert_eq!(output, [1, 2, 1, 2]);
+        assert_eq!(output, [1, 2, 4, 5]);
 
-        let input = vec![1u16, 2, 3, 4, 1, 2, 3, 4];
+        let input = vec![1u16, 2, 3, 4, 5, 6, 7, 8];
         let output = ChannelCountConverter::new(input.into_iter(), 4, 1).collect::<Vec<_>>();
-        assert_eq!(output, [1, 1]);
+        assert_eq!(output, [1, 5]);
     }
 
     #[test]
     fn add_channels() {
-        let input = vec![1u16, 2, 1, 2];
-        let output = ChannelCountConverter::new(input.into_iter(), 2, 3).collect::<Vec<_>>();
-        assert_eq!(output, [1, 2, 2, 1, 2, 2]);
+        let input = vec![1i16, 2, 3, 4];
+        let output = ChannelCountConverter::new(input.into_iter(), 1, 2).collect::<Vec<_>>();
+        assert_eq!(output, [1, 1, 2, 2, 3, 3, 4, 4]);
+
+        let input = vec![1i16, 2];
+        let output = ChannelCountConverter::new(input.into_iter(), 1, 4).collect::<Vec<_>>();
+        assert_eq!(output, [1, 1, 0, 0, 2, 2, 0, 0]);
 
-        let input = vec![1u16, 2, 1, 2];
+        let input = vec![1i16, 2, 3, 4];
         let output = ChannelCountConverter::new(input.into_iter(), 2, 4).collect::<Vec<_>>();
-        assert_eq!(output, [1, 2, 2, 2, 1, 2, 2, 2]);
+        assert_eq!(output, [1, 2, 0, 0, 3, 4, 0, 0]);
+    }
+
+    #[test]
+    fn size_hint() {
```

Comment:
this test seems quite complex and coupled to the implementation. How about something along these lines:

```
fn test(iter, from, to) {
    size_hint = iter.size_hint();
    actual size = iter.count() // from Iterator trait
    assert( is actual size contained in size_hint? )
}
```

If we wanted (might be overkill) we could define a custom Test iterator with a custom uncertain size_hint (I think the min and max size hints are the same for Vec since the len is known).

Reply:
You have a point, but the thing is the tricky/most fallible part of the `size_hint` calculation is when some of the items have already been consumed, so that's the part that is most valuable to test I think.

What about something like this?

```rust
fn test(input: &[i16], from: cpal::ChannelCount, to: cpal::ChannelCount) {
    let count = ChannelCountConverter::new(input.iter().copied(), from, to).count();
    let mut converter = ChannelCountConverter::new(input.iter().copied(), from, to);
    for i in 0..count {
        assert_eq!(converter.size_hint(), (count - i, Some(count - i)));
        converter.next();
    }
}
```

But I am okay with your version if you think this is still overkill.

Reply:
You can make the test function generic and clone the iterator: `fn test<T>(iter: impl Iterator<Item=T> + Clone, from, to)`. Then you can pass in iterators in various states. 

The overkill option would be passing in a custom iterator specifically made for this test. That would look something like this:

```rust
struct TestIter {
  bounds_to_report: (usize, Option<usize>),
  length: usize,
  consumed: usize,
}

impl Iterator for TestIter {
  Item = ();
  fn next(&mut self) -> Self::item {
      if self.consumed >= self.length {
         None
      } else 
         Some(())
      }
  }

  fn size_hint(&self) -> (usize, Option<usize) {
    self.bounds_to_report
  }
}
```

This would allow us to test the behavior when the upper and lower bound are not the same. But I have looked around and no one implementing Iterator does something like this :sweat_smile:, So lets not.

Reply:
I don't see the appeal of the `+ Clone` version. It seems like it'd require a lot of calls to be very useful...

I think these two options are sound:

```rust
fn test_1(input: &[i16], from: cpal::ChannelCount, to: cpal::ChannelCount) {
    let converter = ChannelCountConverter::new(input.iter().copied(), from, to);
    let size_hint = converter.size_hint();
    let count = converter.count();
    assert_eq!(size_hint, (count, Some(count)));
}

fn test_2(input: &[i16], from: cpal::ChannelCount, to: cpal::ChannelCount) {
    let mut converter = ChannelCountConverter::new(input.iter().copied(), from, to);
    let count = converter.clone().count();
    for i in 0..count {
        assert_eq!(converter.size_hint(), (count - i, Some(count - i)));
        converter.next();
    }
}
```

Let me know which one you think is best and I'll commit and push the changes.

Reply:
> I don't see the appeal of the `+ Clone` version. It seems like it'd require a lot of calls to be very useful...

oh seems like that was just me being silly there. I had the count before the size_hint (and thus consuming the iterator) which I fixed using clone. Swapping the order like you do is obv superior :+1: 

The idea of passing in an iterator instead of a slice is that you can pass in partially consumed iterators. You could pass in a `[1,2,3,4].into_iter().skip(3)`. Though the loop works too!

The whole `i in 0..count` and then `count - i`, do you agree with replacing that with `for left_in_iter in (0..count).rev()` and `(left_in_iter, Some(left_in_iter))`?

Reply:
Oh yeah that is better. It'll have to be `(0..=count).rev()` though.
I'm also fixing a bug that caused the size_hint to go up again if calling `.next()` after the iterator was exhausted, and adding one final `assert_eq!` to test for it.